### DISCUSSION TITLE
Regex lookahead requires an external scanner?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1252] [Q&A] [answered]
Just thought I'd get confirmation here before I start cranking out the C++ code. I have a couple places in my parser where I need a regex to capture arbitrary text *except* for a specific string. If the regex encounters that string, it should terminate the match *without capturing that string*, so the string can be matched by other parts of the grammar. For example, my language does nested block comments as follows:
```
(* block comment (* nested block comment *) *)
```
I (somewhat serendipitously) found this is a really good way to parse such comments:
```js
extras: $ => [$.block_comment]
...
block_comment: $ => seq('(*', repeat($.block_comment_text), '*)')
```
Since extras are allowed between sequence elements of `block_comment`, which is itself an extra, nested comments are easily handled without having to muck around with recursive rule definitions. However, this only works if `block_comment_text` does not capture `*)` or `(*`, so the grammar can ensure nested block comment start/end delimiters are matched.

I'd like to write `block_comment_text` as a regex, but it seems like it needs to use two characters of lookahead so it doesn't match/consume `*)` or `(*`. Since tree-sitter is an LR(1) grammar, positive/negative look-ahead is not allowed in regexes. Thus the only way to accomplish this (even though we only need one more character of lookahead!) is in an external scanner. Is my thinking here correct?

ref https://github.com/tlaplus-community/tree-sitter-tlaplus/issues/15

**Comment by ahlinc** [comment, marked as answer]
I think for a complex lookahead the external scanner is the only option available. ~But for your case I suppose it's solvable with a regex.~ The characters lookahead in regexes doesn't allowed for now and I think it's just current tree-sitter limitation in its implementation. But in a limited fashion it's possible to do lookahead with a negated character classes followed some prefix.
Also note `LR(1)` is about grammar tokens and the lookahead of individual characters that are the part of the grammar token can be on any depth.

_I didn't carefully tested bellow grammar example but seems it works pretty good:_
```js
module.exports = grammar({
    name: 'd1252',

    extras: $ => [/\s/, $.block_comment],

    rules: {
        program: $ => repeat(choice('a', 'b', 'c')),
        block_comment: $ => seq(
            '(*',
            optional(
                alias(
                    token(
                        repeat1(/[^*()]|[*][^)]|[(][^*]|[^*][)]|[)][*]/)
                    )
                    , $.block_comment_text
                )
            ),
            '*)'
        ),
    }
});

// ts p <(echo "(*a (* (* 3 *) bbb ccc (* ()* )*( * *(* *) (* 1 (*2*) *) *)*)")
// ts p <(echo "(* ()* )*( *)")
// ts p <(echo "(**)")
```

![Screenshot from 2021-07-10 21-29-15](https://user-images.githubusercontent.com/14666676/125173093-ee6f2380-e1c5-11eb-9d68-e23f683be5c0.png)

> nested comments are easily handled without having to muck around with recursive rule definitions.

@ahelwer  _Interesting finding. Thanks for sharing this!!!_  **- Added:** While it's interesting feature it seems it's impossible to disable it entirely and there may be cases when it's a reason that breaks things and it would be better to have such feature as explicit thing that can be achieved with [explicit recursion](https://github.com/tree-sitter/tree-sitter/issues/931#issuecomment-782905634).

### Edit 1
If it's required to don't consume `\n` inside of tokens and don't have multi line tree-sitter nodes then the above grammar can be adapted to the following:
```js
module.exports = grammar({
    name: 'd1252',

    extras: $ => [/\s/, $.block_comment],

    rules: {
        program: $ => repeat(choice('a', 'b', 'c')),
        block_comment: $ => seq(
            '(*',
            repeat(
                alias(
                    token(
                        repeat1(/[^*()\n]|[*][^)\n]|[(][^*\n]|[^*\n][)]|[)][*]/)
                    )
                    , $.block_comment_text
                )
            ),
            '*)'
        ),
    }
});

// ts p <(echo "(*a (* (* 3 *) bbb ccc (* ()* )*( * *(* *) (* 1 (*2*) *) *)*)")
// ts p <(echo "(* ()* )*( *)")
// ts p <(echo "(**)")

// ts p <(echo -en "(*a (* (* 3 *) bbb\nccc \n ddd eee (* ()* )*( * *(* *) (* 1 (*2*) *) *)*)")
```

![Screenshot from 2021-07-10 21-37-21](https://user-images.githubusercontent.com/14666676/125173304-507c5880-e1c7-11eb-9917-acec054f27bb.png)

### Edit 2
Suggested grammars fail on:
```console
ts p <(echo "(***)")
ts p <(echo "(* **)")
ts p <(echo "(* ****)")
ts p <(echo "(* ******)")
```
To solve this the lookahead on two characters is the must and this can be work in regexes if tree-sitter would support regex lookaheads, internally it has all needed for this the `mark_end()` scanner function that I suppose is used inside of internal scanner also.

**Reply by ahlinc** [reply]
On the above screenshot in the **Edit 1** section the ranges `0:6 - 0:9`, `2:28 - 2:31`, `2:39 - 2:42`, `2:42 - 2:45` seems are tree-sitter issues related to #1200. /cc @maxbrunsfeld

**Reply by ahelwer** [reply]
@ahlinc that regex is a valiant attempt but unfortunately is defeated by the comment `(* text **)`.

Edit: ah just saw your edit, so you noticed already.

**Reply by ahlinc** [reply]
Yeah, if tree-sitter would have the lookahead in regexes that would simplify a lot of cases. But also the problem with it is that the underling `regex-syntax` library also doesn't support the lookahead parsing, so it requires a lot of work to implement it, much more than to write an external scanner for a one token.

**Reply by ahelwer** [reply]
@ahlinc I do think such lookahead might go against the LR(1) parsing design. With regexes, characters are tokens. So allowing more than one character of lookahead is the same as allowing more than one token of lookahead. I do agree it's a lot of work for the user to use an external scanner to implement stuff like "consume until you hit this string of characters" though.

**Reply by HerrmannM** [reply]
@ahelwer I have similar need for a language I'm working one, where start and end of line are significant ('^' and '$' in regexp, which is supported in Lex/Yacc). These kind of "look-around" things are perfectly valid with the LR parsing design, and I think it's actually a feature: the parser deals with tokens, regardless of how they are produced. Deciding which token to emit is the sole responsibility of the lexer. Now, things may get complicated when several lexers are involved (which looks to be the case in TreeSitter), although the current lack of support seems to be more of a library issue than a design issue.

We have an internal parser tool that also support multiple lexers, but they are explicitly defined, as well as when to switch (so, not as usable as TreeSitter). Our lookaheads are handled in the most crude way: push back characters in the lexing stream!
@ahlinc Are you aware if this this is possible with TreeSitter ? i.e. to be able to push back some characters in (or "rewind") the lexing stream?
Edit: I meant to manipulate the lexing stream from the javascript side --- this is available with an external scanner.

**Reply by ahelwer** [reply]
@HerrmannM I ended up implementing it in my external scanner and it works great. See [here](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/5b35152fcca505290c1a99a37f7e7f700a3d267d/src/scanner.cc#L358) and [here](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/5b35152fcca505290c1a99a37f7e7f700a3d267d/src/scanner.cc#L313).

**Reply by HerrmannM** [reply]
Interesting, thanks!

**Reply by ahlinc** [reply]
> Are you aware if this this is possible with TreeSitter ? i.e. to be able to push back some characters in (or "rewind") the lexing stream?

@HerrmannM In Tree-sitter it's possible in the external scanners in a way that you have possibility to decline to consume symbols that you've looked ahead and this is doable by marking such rewind point with the `mark_end()` lexer function. @ahelwer provided great examples how to use it.

**Comment by ahlinc** [comment]
While viewing other similar cases I wasn't able to drop a thought that I've missed something in this case.
So a new grammar version:
```js
module.exports = grammar({
    name: 'd1252',

    extras: $ => [/\s/, $.block_comment],

    rules: {
        program: $ => repeat(choice('a', 'b', 'c')),
        block_comment: $ => seq(
            '(*',
            alias(
                repeat(
                    choice(
                        re(
                            '[^*()\n]',
                            '[^*\n][)]',
                            '[(][^*\n]',
                            '[)][*][^)\n]',
                            '[*][^*)\n]',
                        ),
                        /\*/,
                    ))
                , $.block_comment_text
            ),
            '*)'
        ),
    }
});

function re(regex) {
    if (arguments.length > 1) {
        regex = Array.from(arguments).join('|');
    }
    return {
        type: 'PATTERN',
        value: regex
    };
}

/* Test snippets

ts g && ts p                                                                                \
    <(echo "(*a (* (* 3 *) bbb ccc (* ()* )*( * *(* *) (* 1 (*2*) *) *)*)")                 \
    <(echo -en "(*a (* (* 3 *) bbb\nccc \n ddd eee (* ()* )*( * *(* *) (* 1 (*2*) *) *)*)") \
    <(echo "(* ()* )*( *)")                                                                 \
    <(echo "(**)")                                                                          \
    <(echo "(* **)")                                                                        \
    <(echo "(***)")                                                                         \
    <(echo "(*a*b**)")                                                                      \
    <(echo "(*a(*a(**a**)a*)a*)")

ts g && ts p                                       \
    <(echo "(*a(*a(**a**a*****)a*)a*)")            \
    <(echo "(*a(*a(**a**a****)a*)a*)")             \
    <(echo "(*a(*a(**a**a***)a*)a*)")              \
    <(echo "(*a(*a(**a**a**)a*)a*)")               \
    <(echo "(*a(*a(**a**a*)a*)a*)")                \
    <(echo "(*a(*a(**a)**((a*)a*)a*)")             \
    <(echo "(*a(*a(**a)**((a**)a*)a*)")            \
    <(echo "(*a(*a(**a)**((a***)a*)a*)")           \
    <(echo "(*a(*a(**a)**(*b**)((a***)a*)a*)")     \
    <(echo "(*a(*a(**a)**(*b(**)**)((a***)a*)a*)")

ts g && ts p             \
    <(echo "(***)")      \
    <(echo "(* **)")     \
    <(echo "(* ****)")   \
    <(echo "(* ******)")
*/
```
For the above grammar I can't find any more counter examples.

<details><summary><i>Screenshots:</i></summary>

![image](https://user-images.githubusercontent.com/14666676/130446514-53bd07f1-f48e-4e35-87b3-077714b84dae.png)

![image](https://user-images.githubusercontent.com/14666676/130446416-1fdb5083-f5c6-4b7a-a54e-9a341eae60c7.png)

![image](https://user-images.githubusercontent.com/14666676/130446637-b04273f4-1373-4f50-a299-0b10acd6bb2e.png)

</details>

On the first screenshot it's visible that somehow tree-sitter captures leading spaces for `(*` and `*)` like `" (*"` and `" *)"` but this more looks like a bug and is similar to described in another issue: https://github.com/tree-sitter/tree-sitter/issues/1200#issuecomment-876022936

**Edit 1:**
The above grammar has a one drawback, the `block_comment_text` node has become multi-line, what may be undesirable.
